### ISSUE TITLE
Project ID not automatically determined from key file

### ISSUE THREAD

**Issue by stephenplusplus** [googleapis/nodejs-pubsub#23]
###### Copied from original issue: https://github.com/GoogleCloudPlatform/google-cloud-node/issues/2782

<a href="/victorandree"><img src="https://avatars2.githubusercontent.com/u/52588?s=88&v=4" height=44 width=44 align=left></a>@victorandree<br>December 21, 2017 8:30 AM

Pub/Sub authentication documentation suggests that the project ID will be determined automatically if you provide a `keyFilename` or `GOOGLE_APPLICATION_CREDENTIALS`

> If you wish, you can set an environment variable (GCLOUD_PROJECT) in place of specifying this inline. Or, if you have provided a service account JSON key file as the config.keyFilename property explained above, your project ID will be detected automatically.

Source: https://googlecloudplatform.github.io/google-cloud-node/#/docs/pubsub/0.15.0/guides/authentication

I can't get the "your project ID will be detected automatically" bit to work, and still need to provide `projectId` expressly (either as `GCLOUD_PROJECT` env variable or to the configuration call). This feature isn't really mentioned anywhere else, so maybe it's unsupported? Seems like a good idea though!

This may be related to this issue: https://github.com/GoogleCloudPlatform/google-cloud-node/issues/2576

#### Environment details

  - OS: macOS 10.13.2
  - Node.js version: 8.9.1
  - npm version: 5.6.0
  - @google-cloud/pubsub version: 0.16.0

#### Steps to reproduce

1. Let `pubsub` be configured using only `keyFilename` or `GOOGLE_APPLICATION_CREDENTIALS`
2. Try to subscribe to some topic

Sample code:

```
const pubsub = require('@google-cloud/pubsub')();

const subscription = pubsub.subscription('test-subscription');
subscription.on('message', message => {
  console.log('message received');
  message.ack();
});
```

Error message:

```
events.js:183
      throw er; // Unhandled 'error' event
      ^

Error: Requested project not found or user does not have access to it (project={{projectId}}). Make sure to specify the unique project identifier and not the Google Cloud Console display name.
    at StreamProxy.onConnectionStatus (/code/pubsub/node_modules/@google-cloud/pubsub/src/connection-pool.js:264:21)
    at emitOne (events.js:116:13)
    at StreamProxy.emit (events.js:211:7)
    at ClientDuplexStream.<anonymous> (/code/pubsub/node_modules/google-gax/lib/streaming.js:130:17)
    at emitOne (events.js:116:13)
    at ClientDuplexStream.emit (events.js:211:7)
    at ClientDuplexStream._emitStatusIfDone (/code/pubsub/node_modules/grpc/src/client.js:260:10)
    at ClientDuplexStream._receiveStatus (/code/pubsub/node_modules/grpc/src/client.js:233:8)
    at /code/pubsub/node_modules/grpc/src/client.js:757:12
```

**Comment by stephenplusplus** [comment]
<a href="/victorandree"><img src="https://avatars2.githubusercontent.com/u/52588?s=88&v=4" height=44 width=44 align=left></a>@victorandree<br>December 21, 2017 8:49 AM

I guess the code for Pub/Sub has been moved to https://github.com/googleapis/nodejs-pubsub -- The "Report Issue" links still go here, though...

**Comment by stephenplusplus** [comment]
Sorry for the confusion about the repos.

I haven't been able to reproduce this error. @callmehiphop could you take a look?

**Comment by mscottx88** [comment]
This is still happening as of today.

**Comment by stephenplusplus** [comment]
Any reproduction code I could try?

**Comment by mscottx88** [comment]
This is with `"@google-cloud/pubsub": "^0.16.1"`

```
'use strict';

const Pubsub = require('@google-cloud/pubsub');
const pubsub = new Pubsub();

/**
 * Subscribe to messages for a given topic and subscription.
 *
 * @param {Object} options
 * * maxConcurrentMessages - The maximum number of concurrent messages allowed to process.
 * * subscriptionName - The name of the subscription.
 * * topicName - The name of the topic to subscribe to.
 *
 * @returns {Object}
 * The subscription object is returned.
 */
const subscribeToMessages = ({
  maxConcurrentMessages: maxMessages,
  subscriptionName,
  topicName
}) => {
  return pubsub
    .topic(topicName)
    .subscription(subscriptionName, {
      flowControl: {
        maxMessages
      }
    });
};
```

**Comment by mscottx88** [comment]
Message I am getting: 

```
"Sorry, we cannot connect to Cloud Services without a project ID. You may specify one with an environment variable named "GCLOUD_PROJECT". See https://googlecloudplatform.github.io/google-cloud-node/#/docs/guides/authentication for a detailed guide on creating an authenticated connection."
```

I added environment variable `GCLOUD_PROJECT` as well and still getting this error.

**Comment by mscottx88** [comment]
Downgrading to `0.15.0` seems to resolve the issue...

**Comment by stephenplusplus** [comment]
Thanks, I was able to reproduce this.

@callmehiphop, could you take a look at this? It seems to relate to the frequency that google-auto-auth is being asked for the projectId via `authClient.getProjectId()`. The error is being caught here: https://github.com/stephenplusplus/google-auto-auth/blob/252db3e3226aefa7c3d9f7dce0575e968837f618/index.js#L123, and returned from `authClient.getProjectId()`, but it's not clear why. I'm left thinking maybe there's some throttling/queuing we need to do so we're not calling the auth client so much.

**Comment by stephenplusplus** [comment]
@victorandree we've made a few background changes regarding auth. Could you try again against master?

```sh
$ npm install --save https://github.com/googleapis/nodejs-storage
```

**Comment by victorandree** [comment]
@stephenplusplus Did you want me to try it out with `nodejs-storage` or is that a typo? Trying with `nodejs-pubsub`, I get the same error as @mscottx88 now. Here's my full test environment:

**package.json**
```
{
  "name": "pubsub-test",
  "version": "1.0.0",
  "dependencies": {
    "@google-cloud/pubsub": "git+https://github.com/googleapis/nodejs-pubsub.git"
  }
}
```

**index.js**
```
const pubsub = require('@google-cloud/pubsub')();

const subscription = pubsub.subscription('test-subscription');
subscription.on('message', message => {
  console.log('message received');
  message.ack();
});
```

Running locally (on my MacBook) with no default credentials or default project but with a service account key tied to a project, I get:

> Error: Sorry, we cannot connect to Cloud Services without a project ID. You may specify one with an environment variable named "GCLOUD_PROJECT". See https://googlecloudplatform.github.io/google-cloud-node/#/docs/guides/authentication for a detailed guide on creating an authenticated connection.

Same result if I run in a Docker image with `GOOGLE_APPLICATION_CREDENTIALS` pointed to the service account key mounted as a volume.

**Dockerfile**
```
FROM node:9.5.0
WORKDIR /opt/app
COPY package.json ./
ENV NODE_ENV production
RUN npm install
COPY . .

CMD ["index.js"]
ENTRYPOINT ["node"]
```

**Comment by stephenplusplus** [comment]
@callmehiphop can you try to reproduce?

**Comment by stephenplusplus** [comment]
I believe I found a fix: #74

@victorandree would you mind trying again?

```sh
$ npm install --save git+https://github.com/stephenplusplus/nodejs-pubsub.git#spp--23
```

*double checks for nodejs-storage instead of nodejs-pubsub... nope, we're good :)*

**Comment by victorandree** [comment]
@stephenplusplus Yes, it works! And I verified again that it does not work without this patch so I think you fixed it :)
